src/pkg/services/m365/api/teams.go

```diff
@@ -0,0 +1,161 @@
+	"github.com/microsoftgraph/msgraph-sdk-go/models"
+)
+
+const teamService = "Team"
+
+// ---------------------------------------------------------------------------
+// controller
+// ---------------------------------------------------------------------------
+
+func (c Client) Groups() Groups {
+	return Groups{c}
+}
+
+// On creation of each Microsoft Teams a corrsponding group gets created from them.
+// Most of the information like events, drive and mail info will be fetched directly
+// from groups. So we pull in group and process only the once which are associated with
+// a team for further proccessing of teams.
+
+// Teams is an interface-compliant provider of the client.
+type Groups struct {
+	Client
+}
+
+// GetAll retrieves all groups.
+func (c Groups) GetAll(
+	ctx context.Context,
+	filterTeams bool,
+	errs *fault.Bus,
+) ([]models.Groupable, error) {
+	service, err := c.Service()
+	if err != nil {
+		return nil, err
+	}
+
+	return getGroups(ctx, filterTeams, errs, service)
+}
+
+// GetAll retrieves all groups.
+func getGroups(
+	ctx context.Context,
+	filterTeams bool,
+	errs *fault.Bus,
+	service graph.Servicer,
+) ([]models.Groupable, error) {
+
+	resp, err := service.Client().Groups().Get(ctx, nil)
+	if err != nil {
+		return nil, graph.Wrap(ctx, err, "getting all groups")
+	}
+
+	iter, err := msgraphgocore.NewPageIterator[models.Groupable](
+		resp,
+		service.Adapter(),
+		models.CreateTeamCollectionResponseFromDiscriminatorValue)
+	if err != nil {
+		return nil, graph.Wrap(ctx, err, "creating groups iterator")
+	}
+
+	var (
+		groups = make([]models.Groupable, 0)
+		el     = errs.Local()
+	)
+
+	iterator := func(item models.Groupable) bool {
+		if el.Failure() != nil {
+			return false
+		}
+
+		err := ValidateGroup(item)
+		if err != nil {
+			el.AddRecoverable(ctx, graph.Wrap(ctx, err, "validating groups"))
+		} else {
+			isTeam := IsTeam(item)
+			if !filterTeams || isTeam {
+				groups = append(groups, item)
+			}
+		}
+
+		return true
+	}
+
+	if err := iter.Iterate(ctx, iterator); err != nil {
+		return nil, graph.Wrap(ctx, err, "iterating all groups")
+	}
+
+	return groups, el.Failure()
+}
+
+func IsTeam(g models.Groupable) bool {
+	if g.GetAdditionalData()["resourceProvisioningOptions"] != nil {
+		val, _ := tform.AnyValueToT[[]any]("resourceProvisioningOptions", g.GetAdditionalData())
+		for _, v := range val {
+			s, err := str.AnyToString(v)
+			if err != nil {
+				return false
+			}
+			if s == teamService {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// GetID retrieves team by groupID/teamID.
+func (c Groups) GetByID(
+	ctx context.Context,
+	identifier string,
+) (models.Groupable, error) {
+	service, err := c.Service()
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := service.Client().Groups().ByGroupId(identifier).Get(ctx, nil)
+	if err != nil {
+		return nil, graph.Wrap(ctx, err, "getting group by ID")
+	}
+
+	if err != nil {
+		err := graph.Wrap(ctx, err, "getting teams by id")
```

Comment:
You're not necessarily getting teams here (unless you change the func, as suggested above).
```suggestion
		err := graph.Wrap(ctx, err, "getting groups by id")
```